### PR TITLE
KAFKA-6782: solved the bug of restoration of aborted messages for GlobalStateStore and KGlobalTable

### PR DESCRIPTION
**Pull request by Gitomain**

### CODE REVIEW

**Comment by mjsax** [comment]
\cc @bbejeck @vvcephei

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:230]: Please avoid unnecessary reformatting -- there is more in this PR -- guess, it's a matter of IDE settings?

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java]: I think it would be better, to have two test -- one with EOS enabled and one without.

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java]: Why this change? We don't need the variable `topology`?

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java]: nit: in tests, it better to just declare `throws Exception` -- it does not provide value to spell out the different types in test and introduces "noise"

I realized, that the existing code also does declare different types of exception. Would you mind to clean this up, too?

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java]: nit: add `final` (we try to use `final` whenever possible)

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: nit: formatting -- parameters should be aligned

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: nit: exceptions

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java]: nit: the name is a little miss leading -- aborted message should not be restored...

It also seems, this test only checks if the restore terminates -- I am wondering, if we should also write committed data into the topic and check that the messages got restored? Not sure if necessary.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Couple of comments.

Side note: do we need all those `.gitignore` changes?

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java
Comment (Gitomain) [reply streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:230]: Yes, it's a matter of IDE settings, we don't need these chages.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: nit: formatting - align parameters

**Review by bbejeck — COMMENTED** [review]
Overall looks good just left a minor comment.  I agree with @mjsax about the `.gitignore` changes, can we double check those?

**Comment by mjsax** [comment]
@Gitomain what is the status of this PR -- seems there are still couple of comments you did not address yet.

**Comment by Gitomain** [comment]
Hello, I have deleted all the .gitignore files and added a new test GlobalKTableEOSIntegrationTest to test the EOS case independently.

**Comment by guozhangwang** [comment]
There is a recent commit that moves which package `Consumed` belongs, causing the jenkins to fail:

```
17:02:22 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:26: error: cannot find symbol
17:02:22 import org.apache.kafka.streams.Consumed;
17:02:22                                ^
17:02:22   symbol:   class Consumed
17:02:22   location: package org.apache.kafka.streams
17:02:25 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:114: error: cannot find symbol
17:02:25         globalTable = builder.globalTable(globalTableTopic, Consumed.with(Serdes.Long(), Serdes.String()),
17:02:25                                                             ^
17:02:25   symbol:   variable Consumed
17:02:25   location: class GlobalKTableEOSIntegrationTest
17:02:25 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:118: error: cannot find symbol
17:02:25         final Consumed<String, Long> stringLongConsumed = Consumed.with(Serdes.String(), Serdes.Long());
17:02:25               ^
17:02:25   symbol:   class Consumed
17:02:25   location: class GlobalKTableEOSIntegrationTest
17:02:25 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:118: error: cannot find symbol
17:02:25         final Consumed<String, Long> stringLongConsumed = Consumed.with(Serdes.String(), Serdes.Long());
```

Could you rebase your branch on latest trunk head and merge any conflicts?

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Are those changes intended? Or are they IDE enforced?

**Review by guozhangwang — COMMENTED** [review]
This is a nice catch, thanks @Gitomain . One meta question: what's the motivation of extracting the EOS test into a separate class?

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (Gitomain) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: I think it's a problem of IDE

**Comment by Gitomain** [comment]
For the motivation of extracting the EOS test into a separate class, I think both the EOS enabled and the EOS disabled cases should be tested. And I found it was a little messy to test the 2 cases in the same class. So I extracted the EOS test into a separate class.

**Comment by bbejeck** [comment]
@Gitomain why include `log4j` files for connect and tools?

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java]: nit: formatting

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:114]: with Java 8 support, I think we can remove generics here

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: please address this comment, too

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: do we need a `flush()` here? `abortTransaction()` is a sync call

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java]: this line can be removed

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java]: do we need the timeout? `waitForCondition` already applies a timeout

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:347]: this method should enable EOS by default for this class, shouldn't it?

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java]: we can merge this method into the one above as we always want EOS enabled?

**Review by mjsax — COMMENTED** [review]
Thanks for adding the test!

I think it's ok to split both tests. However, we might want to extract shared code into an abstract super class for both.

Some more follow up comment.

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:75]: Just one meta-comment, now that Streams supports Java 8 do we want to covert the use of anonymous class usage here and other places to lambda expressions?  This comment is very opinionated, and this PR has been going on for a while now, so I wouldn't hold up merging for this.

**Review by bbejeck — COMMENTED** [review]
Just left one last meta-comment.  Overall lgtm, pending addressing @mjsax comments

**Comment by mjsax** [comment]
@Gitomain Code freeze for 2.0 release is on Tuesday (6/12). Can you finish this PR so we can get it into the release? (If not, I might just take it and apply the last changes myself to merged it.) Please let us know.

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (Gitomain) [reply streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:114]: Hello, I can't understand how can we remove generics here, so I didn't modify this part.

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (Gitomain) [reply streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:347]: Thanks for remind, it's done

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (Gitomain) [reply streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:75]: Hello, I know what you mean. But I have seen that all the test codes are still in this style now, I prefer to leave it as it is, what do you think ?

**Comment by Gitomain** [comment]
Hello @mjsax , I have done some modification. But I have no time to extract shared code into an abstract class before 6/12. I think we can finish this PR now.

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:114]: Java8 should be able to infer the type and Java7 was dropped recently. So we want to update the code to Java8 incrementally. You should be able to change this to:
```
Materialized.as(globalStore)
```

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableEOSIntegrationTest.java:75]: As mentioned in the other comment. Java7 was just dropped recently and thus most code is still Java7. We want to incrementally rewrite to Java8. Thus, feel free to update the code accordingly.

**Comment by mjsax** [comment]
The deadline was pushed out by one day. I think we can merge as-is. The code rewrites to Java8 are not too important.

**Review by guozhangwang — APPROVED** [review]

File: .gitignore
Comment (mjsax) [inline .gitignore]: @Gitomain  We can add `*.class` but we need to keep `classes` as some folders are named like this. Can you update the PR? Otherwise I can fix during merging.

**Review by bbejeck — APPROVED** [review]
Thanks for the patch, LGTM

**Comment by mjsax** [comment]
Thanks for the fix @Gitomain 

Merged to `trunk` and cherry-picked to `2.0`, `1.1`, `1.0`, and `0.11.0`.

**Pull request merged**
